File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        if(nums.size() == 0) return {-1,-1};

        if(nums.size() == 1 && target == nums[0]) {
            return {0,0};
        }

        if(nums.size() == 2 && target == nums[0]) {
            if(nums[0] == nums[1]) return {0,1};
            else return {0,0};
        }

        if(nums.size() == 2 && target == nums[1]) {
            return {1,1};
        }

        while(left <= right) {
            int mid = (left + right) / 2;

            if(nums[mid] == target) {

                int i = mid;
                int j = mid;

                // expand left
                while(i-1 >= 0 && nums[i-1] == target) {
                    i--;
                }

                // expand right
                while(j+1 < nums.size() && nums[j+1] == target) {
                    j++;
                }

                return {i, j};
            }
            else if(target < nums[mid]) {
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
        }

        return {-1,-1};
    }
};

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] ans = {-1, -1};

        ans[0] = firstOccurrence(nums, target);
        ans[1] = lastOccurrence(nums, target);

        return ans;
    }

    private int firstOccurrence(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        int index = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target) {
                index = mid;
                right = mid - 1; // go left
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return index;
    }

    private int lastOccurrence(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        int index = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target) {
                index = mid;
                left = mid + 1; // go right
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return index;
    }
}